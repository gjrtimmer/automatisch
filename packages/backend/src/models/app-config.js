import App from './app.js';
import AppAuthClient from './app-auth-client.js';
import Base from './base.js';

class AppConfig extends Base {
  static tableName = 'app_configs';

  static jsonSchema = {
    type: 'object',
    required: ['key'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      key: { type: 'string' },
      canConnect: { type: 'boolean', default: false },
      allowCustomConnection: { type: 'boolean', default: false },
      shared: { type: 'boolean', default: false },
      disabled: { type: 'boolean', default: false },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    appAuthClients: {
      relation: Base.HasManyRelation,
      modelClass: AppAuthClient,
      join: {
        from: 'app_configs.key',
        to: 'app_auth_clients.app_key',
      },
    },
  });

  static get virtualAttributes() {
    return ['canCustomConnect'];
  }

  get canCustomConnect() {
    return !this.disabled && this.allowCustomConnection;
  }

  async getApp() {
    if (!this.key) return null;

    return await App.findOneByKey(this.key);
  }

  async computeCanConnectProperty(oldAppConfig) {
    const appAuthClients = await oldAppConfig.$relatedQuery('appAuthClients');
    const hasSomeActiveAppAuthClients = !!appAuthClients?.some(
      (appAuthClient) => appAuthClient.active
    );
    const shared = this.shared ?? oldAppConfig.shared;
    const disabled = this.disabled ?? oldAppConfig.disabled;
    const active = disabled === false;

    const conditions = [hasSomeActiveAppAuthClients, shared, active];

    const canConnect = conditions.every(Boolean);

    return canConnect;
  }

  async updateCanConnectProperty() {
    const canConnect = await this.computeCanConnectProperty(this);

    return await this.$query().patch({
      canConnect,
    });
  }

  async computeAndAssignCanConnectProperty(oldAppConfig) {
    this.canConnect = await this.computeCanConnectProperty(oldAppConfig);
  }

  async $beforeInsert(queryContext) {
    await super.$beforeInsert(queryContext);

    await this.computeAndAssignCanConnectProperty(this);
  }

  async $beforeUpdate(opt, queryContext) {
    await super.$beforeUpdate(opt, queryContext);

    const oldAppConfig = opt.old;

    await this.computeAndAssignCanConnectProperty(oldAppConfig);
  }
}

export default AppConfig;
