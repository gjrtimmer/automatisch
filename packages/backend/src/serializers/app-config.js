const appConfigSerializer = (appConfig) => {
  return {
    id: appConfig.id,
    key: appConfig.key,
    allowCustomConnection: appConfig.allowCustomConnection,
    shared: appConfig.shared,
    disabled: appConfig.disabled,
    connectionAllowed: appConfig.connectionAllowed,
    canCustomConnect: appConfig.canCustomConnect,
    createdAt: appConfig.createdAt.getTime(),
    updatedAt: appConfig.updatedAt.getTime(),
  };
};

export default appConfigSerializer;
