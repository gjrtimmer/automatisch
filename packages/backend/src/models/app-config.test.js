import { describe, it, expect } from 'vitest';

import Base from './base.js';
import AppConfig from './app-config.js';
import AppAuthClient from './app-auth-client.js';
import { createAppConfig } from '../../test/factories/app-config.js';
import { createAppAuthClient } from '../../test/factories/app-auth-client.js';

describe('AppConfig model', () => {
  it('tableName should return correct name', () => {
    expect(AppConfig.tableName).toBe('app_configs');
  });

  it('jsonSchema should have correct validations', () => {
    expect(AppConfig.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = AppConfig.relationMappings();

    const expectedRelations = {
      appAuthClients: {
        relation: Base.HasManyRelation,
        modelClass: AppAuthClient,
        join: {
          from: 'app_configs.key',
          to: 'app_auth_clients.app_key',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  it('virtualAttributes should return correct properties', () => {
    expect(AppConfig.virtualAttributes).toMatchSnapshot();
  });

  describe('canCustomConnect', () => {
    it('should return true when app is enabled and allows custom connection', async () => {
      const appConfig = await createAppConfig({
        disabled: false,
        allowCustomConnection: true,
      });

      expect(appConfig.canCustomConnect).toBe(true);
    });

    it('should return false when app is disabled', async () => {
      const appConfig = await createAppConfig({
        disabled: true,
        allowCustomConnection: true,
      });

      expect(appConfig.canCustomConnect).toBe(false);
    });

    it(`should return false when app doesn't allow custom connection`, async () => {
      const appConfig = await createAppConfig({
        disabled: false,
        allowCustomConnection: false,
      });

      expect(appConfig.canCustomConnect).toBe(false);
    });
  });

  describe('canConnect', () => {
    it('should return true when app is enabled, shared and allows custom connection', async () => {
      await createAppAuthClient({
        appKey: 'deepl',
        active: true,
      });

      let appConfig = await createAppConfig({
        disabled: false,
        allowCustomConnection: true,
        shared: true,
        key: 'deepl',
      });

      appConfig = await appConfig.$query().withGraphFetched({
        appAuthClients: true,
      });

      expect(appConfig.canConnect).toBe(true);
    });

    it('should return false when app is disabled', async () => {
      const appConfig = await createAppConfig({
        disabled: true,
        allowCustomConnection: true,
      });

      expect(appConfig.canConnect).toBe(false);
    });

    it(`should return false when app doesn't allow custom connection`, async () => {
      const appConfig = await createAppConfig({
        disabled: false,
        allowCustomConnection: false,
      });

      expect(appConfig.canConnect).toBe(false);
    });
  });

  it('getApp should return associated application', async () => {
    const appConfig = await createAppConfig({ key: 'deepl' });

    const app = await appConfig.getApp();

    expect(app.key).toBe('deepl');
  });
});
