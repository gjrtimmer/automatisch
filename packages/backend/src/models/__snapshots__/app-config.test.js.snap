// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppConfig model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "allowCustomConnection": {
      "default": false,
      "type": "boolean",
    },
    "connectionAllowed": {
      "default": false,
      "type": "boolean",
    },
    "createdAt": {
      "type": "string",
    },
    "disabled": {
      "default": false,
      "type": "boolean",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "key": {
      "type": "string",
    },
    "shared": {
      "default": false,
      "type": "boolean",
    },
    "updatedAt": {
      "type": "string",
    },
  },
  "required": [
    "key",
  ],
  "type": "object",
}
`;

exports[`AppConfig model > virtualAttributes should return correct properties 1`] = `
[
  "canCustomConnect",
]
`;
